### PR TITLE
Fix NPE when renaming Rust file

### LINKED ISSUE #615: NPE when renaming a file
```
null
kotlin.KotlinNullPointerException
	at org.rust.lang.core.psi.RustElementFactory.createIdentifier(RustElementFactory.kt:118)
	at org.rust.lang.core.resolve.ref.RustReferenceBase$Companion.doRename(RustReferenceBase.kt:64)
	at org.rust.lang.core.resolve.ref.RustReferenceBase.handleElementRename(RustReferenceBase.kt:47)
	at com.intellij.refactoring.rename.RenameUtil.rename(RenameUtil.java:266)
	at com.intellij.refactoring.rename.RenameUtil.doRenameGenericNamedElement(RenameUtil.java:223)
	at com.intellij.refactoring.rename.RenamePsiElementProcessor.renameElement(RenamePsiElementProcessor.java:60)
	at com.intellij.refactoring.rename.RenameUtil.doRename(RenameUtil.java:187)
	at com.intellij.refactoring.rename.RenameProcessor.performRefactoring(RenameProcessor.java:389)
	at com.intellij.refactoring.BaseRefactoringProcessor$7.run(BaseRefactoringProcessor.java:471)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:906)
	at com.intellij.refactoring.BaseRefactoringProcessor.doRefactoring(BaseRefactoringProcessor.java:453)
	at com.intellij.refactoring.BaseRefactoringProcessor.access$100(BaseRefactoringProcessor.java:76)
	at com.intellij.refactoring.BaseRefactoringProcessor$3.run(BaseRefactoringProcessor.java:293)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:129)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:100)
	at com.intellij.refactoring.BaseRefactoringProcessor.execute(BaseRefactoringProcessor.java:289)
	at com.intellij.refactoring.BaseRefactoringProcessor.doRun(BaseRefactoringProcessor.java:226)
	at com.intellij.refactoring.rename.RenameProcessor.doRun(RenameProcessor.java:122)
	at com.intellij.openapi.fileEditor.impl.NonProjectFileWritingAccessProvider.disableChecksDuring(NonProjectFileWritingAccessProvider.java:188)
	at com.intellij.refactoring.BaseRefactoringProcessor.run(BaseRefactoringProcessor.java:554)
	at com.intellij.refactoring.ui.RefactoringDialog.invokeRefactoring(RefactoringDialog.java:184)
	at com.intellij.refactoring.rename.RenameDialog.performRename(RenameDialog.java:326)
	at com.intellij.refactoring.rename.RenameDialog.doAction(RenameDialog.java:304)
	at com.intellij.refactoring.rename.RenameWithOptionalReferencesDialog.doAction(RenameWithOptionalReferencesDialog.java:60)
	at com.intellij.refactoring.ui.RefactoringDialog.doRefactorAction(RefactoringDialog.java:89)
	at com.intellij.refactoring.ui.RefactoringDialog$RefactorAction.actionPerformed(RefactoringDialog.java:161)
	at javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:2022)
	at javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2348)
	at javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:402)
	at javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:259)
	at javax.swing.AbstractButton.doClick(AbstractButton.java:376)
	at javax.swing.plaf.basic.BasicRootPaneUI$Actions.actionPerformed(BasicRootPaneUI.java:208)
	at javax.swing.SwingUtilities.notifyAction(SwingUtilities.java:1663)
	at javax.swing.JComponent.processKeyBinding(JComponent.java:2882)
	at javax.swing.KeyboardManager.fireBinding(KeyboardManager.java:307)
	at javax.swing.KeyboardManager.fireKeyboardAction(KeyboardManager.java:250)
	at javax.swing.JComponent.processKeyBindingsForAllComponents(JComponent.java:2974)
	at javax.swing.JComponent.processKeyBindings(JComponent.java:2966)
	at javax.swing.JComponent.processKeyEvent(JComponent.java:2845)
	at java.awt.Component.processEvent(Component.java:6310)
	at java.awt.Container.processEvent(Container.java:2236)
	at java.awt.Component.dispatchEventImpl(Component.java:4889)
	at java.awt.Container.dispatchEventImpl(Container.java:2294)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.KeyboardFocusManager.redispatchEvent(KeyboardFocusManager.java:1954)
	at java.awt.DefaultKeyboardFocusManager.dispatchKeyEvent(DefaultKeyboardFocusManager.java:806)
	at java.awt.DefaultKeyboardFocusManager.preDispatchKeyEvent(DefaultKeyboardFocusManager.java:1074)
	at java.awt.DefaultKeyboardFocusManager.typeAheadAssertions(DefaultKeyboardFocusManager.java:945)
	at java.awt.DefaultKeyboardFocusManager.dispatchEvent(DefaultKeyboardFocusManager.java:771)
	at java.awt.Component.dispatchEventImpl(Component.java:4760)
	at java.awt.Container.dispatchEventImpl(Container.java:2294)
	at java.awt.Window.dispatchEventImpl(Window.java:2746)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.awt.EventQueue$4.run(EventQueue.java:731)
	at java.awt.EventQueue$4.run(EventQueue.java:729)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:863)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:678)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:385)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:109)
	at java.awt.WaitDispatchSupport$2.run(WaitDispatchSupport.java:184)
	at java.awt.WaitDispatchSupport$4.run(WaitDispatchSupport.java:229)
	at java.awt.WaitDispatchSupport$4.run(WaitDispatchSupport.java:227)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.awt.WaitDispatchSupport.enter(WaitDispatchSupport.java:227)
	at java.awt.Dialog.show(Dialog.java:1084)
	at com.intellij.openapi.ui.impl.DialogWrapperPeerImpl$MyDialog.show(DialogWrapperPeerImpl.java:778)
	at com.intellij.openapi.ui.impl.DialogWrapperPeerImpl.show(DialogWrapperPeerImpl.java:455)
	at com.intellij.openapi.ui.DialogWrapper.invokeShow(DialogWrapper.java:1665)
	at com.intellij.openapi.ui.DialogWrapper.show(DialogWrapper.java:1607)
	at com.intellij.refactoring.ui.RefactoringDialog.access$101(RefactoringDialog.java:38)
	at com.intellij.refactoring.ui.RefactoringDialog.lambda$show$1(RefactoringDialog.java:189)
	at com.intellij.openapi.project.DumbPermissionServiceImpl.allowStartingDumbModeInside(DumbPermissionServiceImpl.java:37)
	at com.intellij.openapi.project.DumbService.allowStartingDumbModeInside(DumbService.java:283)
	at com.intellij.refactoring.ui.RefactoringDialog.show(RefactoringDialog.java:189)
	at com.intellij.refactoring.rename.PsiElementRenameHandler.rename(PsiElementRenameHandler.java:198)
	at com.intellij.refactoring.rename.PsiElementRenameHandler.rename(PsiElementRenameHandler.java:169)
	at com.intellij.refactoring.rename.PsiElementRenameHandler.invoke(PsiElementRenameHandler.java:119)
	at com.intellij.refactoring.rename.PsiElementRenameHandler.invoke(PsiElementRenameHandler.java:94)
	at com.intellij.refactoring.actions.BaseRefactoringAction.actionPerformed(BaseRefactoringAction.java:126)
	at com.intellij.openapi.actionSystem.ex.ActionUtil$1.run(ActionUtil.java:197)
	at com.intellij.openapi.application.TransactionGuardImpl.runSyncTransaction(TransactionGuardImpl.java:88)
	at com.intellij.openapi.application.TransactionGuardImpl.submitTransactionAndWait(TransactionGuardImpl.java:156)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.performActionDumbAware(ActionUtil.java:211)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher$1.performAction(IdeKeyEventDispatcher.java:587)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.lambda$processAction$2(IdeKeyEventDispatcher.java:634)
	at com.intellij.openapi.application.TransactionGuardImpl.performUserActivity(TransactionGuardImpl.java:199)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.processAction(IdeKeyEventDispatcher.java:633)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.inInitState(IdeKeyEventDispatcher.java:477)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.dispatchKeyEvent(IdeKeyEventDispatcher.java:211)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:677)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:385)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

plugin version 0.0.1.903, Idea 2016.2.2

It is not a big problem, I can rename a file in terminal, if it does not tell you anything, please close the issue.

### PR DESCRIPTION
**Pull request by zjhmale**
Fixes #615

### CODE REVIEW

**Comment by matklad** [comment]
In the long run we need to do something more clever here, like fixing all the `mod decl`. Can you add a test for the NPE?

**Comment by zjhmale** [comment]
@matklad yeah, i admit this is just a quick patch 😂 
>In the long run we need to do something more clever here, like fixing all the mod decl

i back to this again and find it is quit difficult to add test for this since there is no concrete pattern to trigger this issue.

snapshot

the contents inside `foo.rs` and `hahaha.rs` are exactly the same

```rust
pub enum foo {
}

fn bar() {
    ((3 * 3));
}
```

and when renaming `hahaha.rs` it will trigger the NPE where as renaming `foo.rs` it just don't enter into [this function](https://github.com/intellij-rust/intellij-rust/blob/master/src/main/kotlin/org/rust/lang/core/resolve/ref/RustReferenceBase.kt) as expected.

>Can you add a test for the NPE?

BTW, is there anyway to test renaming file? i can only find renaming function name or other PsiElement cases.

**Comment by matklad** [comment]
>and when renaming hahaha.rs it will trigger the NPE where as renaming foo.rs it just don't enter into this function as expected.

I am not sure, but perhaps you have `mod hahaha;` but not `mod foo;`? 

>BTW, is there anyway to test renaming file? i can only find renaming function name or other PsiElement cases.

`RustFile` is a PsiElement, so the usual renaming should work just fine, doesn't' ir?

**Comment by zjhmale** [comment]
exactly, thx! i just forgot to search around the whole project 😅 
>I am not sure, but perhaps you have mod hahaha; but not mod foo;?

yeah, i know PsiFile is also a PsiElement. i mean, when mock renaming other PsiElement like function name or something like that there will be a caret to index the position of that PsiElement but how about PsiFile? anyway i'll try to figure it out.
>RustFile is a PsiElement, so the usual renaming should work just fine, doesn't' it?

File: src/test/kotlin/org/rust/ide/refactoring/RenameTest.kt [hunk 1]
```diff
@@ -11,7 +11,12 @@ class RenameTest: RustTestCaseBase() {
     private fun doTest(name: String = "spam") {
         myFixture.configureByFile(fileName)
         myFixture.renameElementAtCaret(name)
-        myFixture.checkResultByFile(fileName, fileName.replace(".rs", "_after.rs"), false);
+        myFixture.checkResultByFile(fileName, fileName.replace(".rs", "_after.rs"), false)
+    }
+
+    fun testRenameFile() = checkByDirectory {
```
Comment (zjhmale) [inline src/test/kotlin/org/rust/ide/refactoring/RenameTest.kt]: @matklad the NPE test is added, as the reason to spur the NPE is that if we rename `foo.rs` if we have usage of `foo` module in other file.

**Comment by matklad** [comment]
Thanks!

**Comment by adelarsq** [comment]
A little off the topic. Which plugin are you guys using to show specific icons for each filetype on Intellij (for Cargo.lock, Cargo.toml, ...)?

**Comment by matklad** [comment]
@adelarsq this one: https://github.com/intellij-rust/intellij-toml but it doesn't have much stuff in it yet.

**Comment by adelarsq** [comment]
@matklad I'm already using 😄. I mean the icons for each type on the Project tree. For me the only way was using [AfterglowIntelliJ](https://github.com/JustSid/AfterglowIntelliJ) theme, that have icons only for *.rs files. I'm using Intellij Community.

**Comment by matklad** [comment]
@adelarsq Hm, that's strange. Each file type recognized by IDEA should have a specific icon. Am I correct that the issue you face is as follows:

* Rust/Toml files use an icon of plain text file
* Java/Kotlin/JavaScript etc. use specific icons

**Comment by mkaput** [comment]
@adelarsq have you set up your project correctly? We set icons for Cargo.toml and Cargo.lock only when appropriate - when they are in project with Rust toolchain set up

**Comment by adelarsq** [comment]
@matklad exactly. This is a screenshot for a simple project using the last plugin version:

image

@mkaput yes. Just the icons don't work properly. I was thinking that was no support for icons, until today that I see the screenshot from this issue and realized that something is wrong.

**Comment by mkaput** [comment]
Hmm seems like afterglow plugin issue. Maybe it's handling more popular plugins especially like .ignore?

**Comment by adelarsq** [comment]
@mkaput no. I have disabled afterglow to show the problem. I still don't have gitignore plugin installed.

**Comment by zjhmale** [comment]
@adelarsq there exists [rust icons](https://github.com/JustSid/AfterglowIntelliJ/tree/master/resources/file-icons), maybe you can open an issue there.

**Comment by adelarsq** [comment]
@zjhmale yes, but only if I use this theme. Isn't transparency. I'm using the background feature from Intellij. To access search for "Set Background Image" in the "Find Action..." (Ctrl+Shift+A)

**Comment by zjhmale** [comment]
@adelarsq no worries, i just missed the transparency feature of editors like vim or emacs sooooo much 😅

**Pull request merged**